spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
+true in case the header is within its lite client trusted period.
+```verify``` function is capturing basic header verification, i.e., it ensures that the header is signed by more than 2/3 of the voting power of the corresponding validator set.
 
-**Auxiliary Functions.** We will use the function `votingpower_in(V1,V2)` to compute the voting power the validators in set V1 have according to their voting power in set V2;
-we will write `totalVotingPower(V)` for `votingpower_in(V,V)`, which returns the total voting power in V.
-We further use the function `signers(Commit)` that returns the set of validators that signed the Commit.
+```go
+  // return true if header is within its lite client trusted period; otherwise it returns false
+  func isWithinTrustedPeriod(h) bool {
+     return h.Header.bfttime + LITE_CLIENT_TRUSTED_PERIOD > now
+  }
 
-**CheckSupport.** The following function checks whether we can trust the header h2 based on header h1 following the trusting period method.
+  // return true if header is correctly signed by 2/3+ voting power in the corresponding
+  // validator set; otherwise false. Additional checks should be done in the implementation
+  // to ensure header is well formed.
+  func verify(h) bool {
+    vp_all := totalVotingPower(h.Header.V) // total sum of voting power of validators in h
+    return votingpower_in(signers(h.Commit),h.Header.V) > 2/3 * vp_all
+  }
 
-```go
-  func CheckSupport(h1,h2,trustlevel) bool {
-    if h1.Header.bfttime + tp < now { // Observation 1
-      return false // old header was once trusted but it is expired
-    }
-    vp_all := totalVotingPower(h1.Header.NextV)
-      // total sum of voting power of validators in h2
-
-    if h2.Header.height == h1.Header.height + 1 {
-      // specific check for adjacent headers; everything must be
-      // properly signed.
-      // also check that h2.Header.V == h1.Header.NextV
-      // Plus the following check that 2/3 of the voting power
-      // in h1 signed h2
-      return (votingpower_in(signers(h2.Commit),h1.Header.NextV) >
-              2/3 * vp_all)
-        // signing validators are more than two third in h1.
-    }
+  // Captures skipping condition. h1 and h2 have already passed basic validation
+  // (function `verify`).
+  // Returns nil in case h2 can be trusted based on h1, otherwise returns error.
+  // ErrHeaderExpired is used when h1 has expired with respect to lite client trusted period,
+  // ErrInvalidAdjacentHeaders when that adjacent headers are not consistent and
+  // ErrTooMuchChange when there is not enough intersection between validator sets to have
+  // skipping condition true.
+  func CheckSupport(h1,h2,trustThreshold) error {
+    assume h1.Header.Height < h2.header.Height and
+           h1.Header.bfttime < h2.Header.bfttime and
+           h2.Header.bfttime < now
```

Comment:
This `assume` this could provide more guidance for an implementer: Is it like an assert (would panic if called with the assumptions violated), or should it simply return an error? Or is this mere documentation? I think a spec should be explicit on how to (should, must, may, shall) handle a violation of these assumptions.

Reply:
Very good point. I think this is assert (maybe we should actually change it to assert) and say that if this basic assumptions do not hold, we panic.